### DISCUSSION TITLE
Parallel Processing with Pool

### DISCUSSION THREAD

**Discussion by sencer4898** [ourownstory/neural_prophet#534] [Q&A - get help using NeuralProphet] [answered]
Hi. I am trying to run Neural Prophet model for different, similar entities. It is parallelized, however model progress is too slow/close to none. This is the part where I do the parallelizing:
![prob1](https://user-images.githubusercontent.com/41899243/153001315-78c21dc0-fce4-4b38-aff0-c168678d5be9.png)

model_array is the array that keeps models for each entity. Here is the part where the progress gets stuck:
![prob2](https://user-images.githubusercontent.com/41899243/153001455-d22e51af-5eaf-4e0b-b027-af6f6de5524b.png)

It progresses at times (discrete time progress) but it is 10x slower than the serial processing, therefore the problem. Also, it uses all of the cores of local computer (88 cores), even though I specify it to be 4. What could be the problem?

**Comment by ourownstory** [comment, marked as answer]
Hi @sencer4898 
We have also observed some odd behavior depending which type of multiprocessing function we used. (In particular, combining ThreadPools and regular Pools lead to odd behavior.)
The one setup that worked reliably on all our machines thus far was to use `map_async`
You can see two examples here: https://github.com/ourownstory/neural_prophet/blob/da317175f45c093dedb7e9603f6b1bb15e654b9e/neuralprophet/benchmark.py#L435
and here: 
https://github.com/ourownstory/neural_prophet/blob/da317175f45c093dedb7e9603f6b1bb15e654b9e/neuralprophet/benchmark.py#L516

Please let me know if this approach works for you and share your updated code snippet. 
Thank you,
Oskar

**Reply by sencer4898** [reply]
Hi @ourownstory . Thanks for the quick reply, much appreciated.

I tried to run the code the way you suggested. However, the problem still continues. Its been 10 minutes and the script is still at this stage shown below, with no progress at all.
<img width="612" alt="model_stuck" src="https://user-images.githubusercontent.com/41899243/153143565-6eca2bb5-a808-4f96-adb9-92f2b629a092.png">

These are the main and callback functions:
<img width="609" alt="functions" src="https://user-images.githubusercontent.com/41899243/153143710-1825b7cb-aa1a-432d-8a9a-41dce7730c44.png">

This is the array that I keep all the models (different model for each entity):
<img width="443" alt="model_array1" src="https://user-images.githubusercontent.com/41899243/153143823-3d4187ac-2bfa-46b4-bcba-70d2ec9fccbe.png">

Also, on a side note, I am working on the local server of my company. This is the only configuration attribute that I set up:
`os.environ['NUMEXPR_MAX_THREADS'] = '40'`

I can give additional information if you need any. Thanks a lot.

Mehmet

**Reply by sencer4898** [reply]
I noticed that I return a variable called 'i', which doesn't exist, in the function. However, it should not be an issue because I couldn't complete the model fitting stage anyways.

**Reply by ourownstory** [reply]
Hi Mehmet  I am not sure why the same method is not working for you. I wish I understood mp better, to help you better. 
As of now, my best guess is that in my use of mp, I instantiate the model in the parallelized function, while you instantiate it before. 
I suggest creating a minimal setup to re-create the problem with the least amount of confounding factors to better identify the root cause.

Please do let us know if you find a solution. I am sure other people will appreciate it!
